packages/google_sign_in/lib/src/oauth2.dart

```diff
@@ -0,0 +1,276 @@
+      expiresIn: Duration(seconds: json['expires_in'] as int),
+      refreshToken: json['refresh_token'] as String?,
+      scope: json['scope'] != null
+          ? (json['scope'] as String).split(' ').toList()
+          : null,
+      idToken: json['id_token'] as String,
+    );
+  }
+
+  /// The access token issued by the authorization server.
+  final String accessToken;
+
+  /// The type of the token issued.
+  ///
+  /// In the context of Google sign-in, this type is always 'Bearer'.
+  final String tokenType;
+
+  /// The lifetime of the [accessToken].
+  final Duration expiresIn;
+
+  /// The token that can be used to obtain new access tokens when they are expired.
+  final String? refreshToken;
+
+  /// The API scope of the access token.
+  final List<String> scope;
+
+  /// The token issued by the Google authorization server that holds Google
+  /// account information.
+  final String idToken;
+}
+
+/// OAuth 2.0 client that handles Device Authorization Grant.
+/// See:
+///  - [Google SignIn guide for limited input devices](https://developers.google.com/identity/gsi/web/guides/devices)
+///  - [Google OAuth 2.0 guide for limited input devices](https://developers.google.com/identity/protocols/oauth2/limited-input-device)
+///  - [OAuth 2.0 Device Authorization Grant spec](https://datatracker.ietf.org/doc/html/rfc8628)
+///  - [OpenID Connect spec](https://openid.net/specs/openid-connect-core-1_0.html)
+class DeviceAuthClient {
+  /// Creates an instance of [DeviceAuthClient].
+  DeviceAuthClient({
+    required this.authorizationEndPoint,
+    required this.tokenEndPoint,
+    required this.revokeEndPoint,
+    http.Client? httpClient,
+  }) : _httpClient = httpClient ?? http.Client();
+
+  /// The server endpoint that returns authroization grant when user grants access.
+  final Uri authorizationEndPoint;
+
+  /// The server endpoint that returns access token when given authorization grant.
+  final Uri tokenEndPoint;
+
+  /// The server endpoint that revokes future access to issued tokens.
+  final Uri revokeEndPoint;
+
+  final http.Client _httpClient;
+
+  bool _isPolling = false;
+
+  /// Checks whether polling started by by [pollToken] is in progress.
+  bool get isPolling => _isPolling;
+
+  /// Stops poll request started by [pollToken].
+  void cancelPollToken() => _isPolling = false;
+
+  /// Requests authroization grant from [authorizationEndPoint].
+  Future<AuthorizationResponse> requestAuthorization(
+      String clientId, List<String> scope) async {
+    final Map<String, String> body = <String, String>{
+      'client_id': clientId,
+      'scope': scope.join(' '),
+    };
+
+    final http.Response response =
+        await _httpClient.post(authorizationEndPoint, body: body);
+
+    final Map<String, dynamic> jsonResponse =
+        convert.jsonDecode(response.body) as Map<String, dynamic>;
+    if (response.statusCode != 200) {
+      _handleErrorResponse(jsonResponse);
```

Comment:
Is `response.body` still a valid JSON document even if `response.statusCode` is non-OK (200)?

I think you should also check for the status code 400 according to https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-31#section-5.2.

Reply:
Done!